File: Wirtualny_Swiat_Java/Projekt_PO_Java/src/model/Organisms/Cyber_Sheep.java
package model.Organisms;

import model.Animal;
import model.Organism;
import model.Point;
import model.World;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;


public class Cyber_Sheep extends Animal {
    private Color cyberColor = Color.CYAN;

    public Cyber_Sheep(World world, Point position){
        super(world, position, 11, 4);
    }

    @Override
    public Color getColor(){
        return cyberColor;
    }

    @Override
    public String getNazwa(){
        return "Cyber_Owca";
    }

    @Override
    public void Action(){
        Point currentPos = getPosition();
        List<Point> barszcze = new ArrayList<>();
        List<Organism> organisms = world.getOrganisms();
        Point closest = null;
        double minDist = 100000;

        for (Organism o : organisms){
            if (o instanceof BarszczSosnowskiego){
                barszcze.add(o.getPosition());
            }
        }

        for(Point barszcz : barszcze){
            double dist = distance(this.getPosition(), barszcz);
            if (dist < minDist){
                minDist = dist;
                closest = new Point(barszcz.getX(), barszcz.getY());
            }
        }

        if (closest != null){
            int dx = Integer.compare(closest.getX(), currentPos.getX());
            int dy = Integer.compare(closest.getY(), currentPos.getY());

            Point newPos = new Point(currentPos.getX() + dx, currentPos.getY() + dy);

            world.tryToMoveOrganism(this, newPos);
        }else {
            super.Action();
        }
    }

    private double distance(Point a, Point b){
        int dx = a.getX() - b.getX();
        int dy = a.getY() - b.getY();
        return Math.sqrt(dx*dx + dy*dy);
    }

    @Override
    public Organism makeNewOrganism(Point position){
        return new Cyber_Sheep(world, position);
    }
}

File: WirtualnySwiatPython/Animals/CyberSheep.py
import math

from Plants.BarszczSosnowskiego import BarszczSosnowskiego
from animal import Animal
from config import BLUE
from point import Point


class CyberSheep(Animal):
    def __init__(self, world, position):
        super().__init__(world, position, power=11, initiative=4)
        self.color = BLUE

    def getColor(self):
        return self.color

    def getNazwa(self):
        return "CyberOwca"

    def makeNewOrganism(self, position):
        return CyberSheep(self.world, position)

    def Action(self):
        position = self.getPosition()
        barszcze = []
        minDist = 100000000000000000000
        closest = None

        for organism in self.world.organisms:
            if isinstance(organism, BarszczSosnowskiego):
                barszcze.append(organism.getPosition())

        for barszcz in barszcze:
            dist = self.distance(position, barszcz)
            if (dist < minDist):
                minDist = dist
                closest = Point(barszcz.GetX(), barszcz.GetY())

        if closest is not None:
            self.increaseAge()
            dx = (closest.GetX() > position.GetX()) - (closest.GetX() < position.GetX())
            dy = (closest.GetY() > position.GetY()) - (closest.GetY() < position.GetY())

            new_pos = Point(position.GetX() + dx, position.GetY() + dy)

            self.world.tryToMoveOrganism(self, new_pos)
        else:
            super().Action()

    def distance(self, A, B):
        dx = A.GetX() - B.GetX()
        dy = A.GetY() - B.GetY()
        return math.sqrt(dx*dx + dy*dy)